src/cairo/bootloader.md

```diff
@@ -1 +1,199 @@
 # Bootloader
+
+The bootloader is a Cairo program that can prove the execution of multiple Cairo programs by simply running the programs inside the bootloader. Since the Cairo program to be proved can be any arbitrary program, it can also be a Cairo verifier program, which means that the bootloader also supports creating a proof that proves the validity of a bootloader proof. Building on this concept, the bootloader allows creating a proof of nested Cairo programs, or a tree-like structure made up of Cairo programs and bootloader proofs.
+
+If this sounds too abstract, don't worry--this doc will attempt to explain in detail how the bootloader works and the two use cases mentioned above. (Note: this is based on the [Cairo v0.13.1 version of the bootloader](https://github.com/starkware-libs/cairo-lang/tree/v0.13.1/src/starkware/cairo/bootloaders))
+
+## A short primer on how CairoVM works
+
+Before diving into the details about the bootloader, let’s go over how CairoVM works. CairoVM is a virtual machine that can execute a Cairo program (technically, a Cairo program compiled to a Cairo-specific set of instructions) and in the end produce a memory vector and execution trace table that is populated throughout the execution of the program.
+
+More specifically, the generated memory can be thought of as a vector that is divided into segments that have different functionalities (e.g. program instructions, execution, output). The execution trace is a table with 3 columns each representing a register value and rows that each represent the value of the registers over the execution of the program. This means that the number of rows is equal to the number of steps that the CairoVM runs, given that 1 step equals 1 instruction call.
```

Comment:
What's the "generated memory"?

> The execution trace is a table with 3 columns each representing a register value and rows that each represent the value of the registers over the execution of the program.

It's slightly confusing to mix the execution trace with the memory. Presumably the registers are pointers to the memory (like the program counter).

Reply:
> given that 1 step equals 1 instruction call.

Is this saying that the CairoVM is a single cycle processor?

Reply:
> What's the "generated memory"?
Changed to "fully assigned memory"

> Is this saying that the CairoVM is a single cycle processor?
Yes